python-threatexchange/threatexchange/cli/cli_state.py

```diff
@@ -0,0 +1,198 @@
+#!/usr/bin/env python
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+"""
+A wrapper around loading and storing ThreatExchange data from files.
+
+There are a few categories of state that this wraps:
+  1. Checkpoints - state about previous fetches
+  2. Collaboration Indicator Dumps - Raw output from threat_updates
+  3. Index state - serializations of indexes for SignalType
+"""
+
+from enum import Enum
+import json
+import pathlib
+import typing as t
+import dataclasses
+import logging
+
+import dacite
+
+from threatexchange.signal_type.index import SignalTypeIndex
+from threatexchange.signal_type.signal_base import SignalType
+from threatexchange.cli.exceptions import CommandError
+from threatexchange.fetcher.collab_config import CollaborationConfigBase
+from threatexchange.fetcher.fetch_state import (
+    FetchCheckpointBase,
+    FetchedSignalMetadata,
+)
+from threatexchange.fetcher.simple import state as simple_state
+from threatexchange.fetcher.fetch_api import SignalExchangeAPI
+from threatexchange.signal_type import signal_base
+from threatexchange.signal_type import index
+
+
+class FetchCheckpoint(t.NamedTuple):
+    last_full_fetch: float
+    last_fetch: float
+
+
+class Dataset:
+
+    EXTENSION = ".te"
+
+
+class CliIndexStore:
+
+    FILE_EXTENSION = ".index"
+
+    def __init__(self, indice_dir: pathlib.Path) -> None:
+        self.dir = indice_dir
+
+    def get_available(self) -> t.List[str]:
+        return [
+            str(f)[-len(self.FILE_EXTENSION)]
+            for f in self.dir.glob(f"*{self.FILE_EXTENSION}")
+        ]
+
+    def clear(
+        self, only_types: t.Optional[t.Iterable[t.Type[SignalType]]] = None
```

Comment:
nit: for collections, default to an empty list works just as good as t.Optional[t.Iterable[...]]. 

You'll not need to add guard clauses.

Reply:
This is a deliberate choice to avoid confusing with "I choose to restrict to nothing" versus "I choose not to restrict". I believe the change you are suggesting will cause an actual behavior change due to the suggestion.